File: test/domain/measurement.test.ts
import { describe, expect, it } from "vitest";

import {
  latestByKind,
  measurementPeriodLabels,
  selectReadingsByWeightAnchor,
  selectWeightByPeriod,
} from "../../src/domain/measurement.js";
import type { MeasurementReading } from "../../src/domain/measurement.js";

describe("measurement domain contract", () => {
  it("maps normalized periods to spreadsheet labels", () => {
    expect(measurementPeriodLabels).toEqual({
      morning: "朝",
      evening: "夜",
    });
  });

  it("selects the latest reading for each measurement kind", () => {
    const readings: MeasurementReading[] = [
      {
        kind: "weight",
        value: 70.1,
        unit: "kg",
        measuredAt: "2026-06-18T07:00:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "weight",
        value: 70.3,
        unit: "kg",
        measuredAt: "2026-06-18T07:05:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "pulse",
        value: 64,
        unit: "bpm",
        measuredAt: "2026-06-18T07:03:00.000Z",
        source: "apple_health_export",
      },
    ];

    const latest = latestByKind(readings);

    expect(latest.get("weight")?.value).toBe(70.3);
    expect(latest.get("pulse")?.value).toBe(64);
  });

  it("selects weight by period", () => {
    const readings: MeasurementReading[] = [
      {
        kind: "weight",
        value: 70.1,
        unit: "kg",
        measuredAt: "2026-06-18T07:00:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "weight",
        value: 70.3,
        unit: "kg",
        measuredAt: "2026-06-18T07:05:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "pulse",
        value: 64,
        unit: "bpm",
        measuredAt: "2026-06-18T07:03:00.000Z",
        source: "apple_health_export",
      },
    ];

    expect(selectWeightByPeriod(readings, "morning")?.value).toBe(70.1);
    expect(selectWeightByPeriod(readings, "evening")?.value).toBe(70.3);
    expect(selectWeightByPeriod(readings.slice(2), "morning")).toBeUndefined();
  });

  it("selects readings around the weight anchor", () => {
    const readings: MeasurementReading[] = [
      {
        kind: "weight",
        value: 70.1,
        unit: "kg",
        measuredAt: "2026-06-18T07:00:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "weight",
        value: 70.3,
        unit: "kg",
        measuredAt: "2026-06-18T07:10:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "body_temperature",
        value: 36.4,
        unit: "celsius",
        measuredAt: "2026-06-18T06:55:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "body_temperature",
        value: 36.7,
        unit: "celsius",
        measuredAt: "2026-06-18T07:03:00.000Z",
        source: "apple_health_export",
      },
      {
        kind: "pulse",
        value: 64,
        unit: "bpm",
        measuredAt: "2026-06-18T07:11:00.000Z",
        source: "apple_health_export",
      },
    ];

    const morning = selectReadingsByWeightAnchor(readings, "morning");
    const evening = selectReadingsByWeightAnchor(readings, "evening");

    expect(morning.get("weight")?.value).toBe(70.1);
    expect(morning.get("body_temperature")?.value).toBe(36.7);
    expect(morning.get("pulse")?.value).toBe(64);
    expect(evening.get("weight")?.value).toBe(70.3);
    expect(evening.get("body_temperature")?.value).toBe(36.7);
    expect(evening.get("pulse")?.value).toBe(64);
    expect(selectReadingsByWeightAnchor(readings.slice(2), "morning").size).toBe(
      0,
    );
  });
});

File: internal/domain/measurement_test.go
package domain

import "testing"

func TestMeasurementDomainContract(t *testing.T) {
	t.Run("period labels", func(t *testing.T) {
		if got, want := PeriodLabels[PeriodMorning], "朝"; got != want {
			t.Fatalf("morning label = %q, want %q", got, want)
		}
		if got, want := PeriodLabels[PeriodEvening], "夜"; got != want {
			t.Fatalf("evening label = %q, want %q", got, want)
		}
	})

	readings := []MeasurementReading{
		{Kind: KindWeight, Value: 70.1, Unit: UnitKg, MeasuredAt: "2026-06-18T07:00:00.000Z", Source: "apple_health_export"},
		{Kind: KindWeight, Value: 70.3, Unit: UnitKg, MeasuredAt: "2026-06-18T07:05:00.000Z", Source: "apple_health_export"},
		{Kind: KindPulse, Value: 64, Unit: UnitBPM, MeasuredAt: "2026-06-18T07:03:00.000Z", Source: "apple_health_export"},
	}

	t.Run("latest by kind", func(t *testing.T) {
		latest := LatestByKind(readings)
		if got := latest[KindWeight].Value; got != 70.3 {
			t.Fatalf("latest weight = %v, want 70.3", got)
		}
		if got := latest[KindPulse].Value; got != 64 {
			t.Fatalf("latest pulse = %v, want 64", got)
		}
	})

	t.Run("weight by period", func(t *testing.T) {
		if got := SelectWeightByPeriod(readings, PeriodMorning); got == nil || got.Value != 70.1 {
			t.Fatalf("morning weight = %#v, want 70.1", got)
		}
		if got := SelectWeightByPeriod(readings, PeriodEvening); got == nil || got.Value != 70.3 {
			t.Fatalf("evening weight = %#v, want 70.3", got)
		}
		if got := SelectWeightByPeriod(readings[2:], PeriodMorning); got != nil {
			t.Fatalf("weight without weight readings = %#v, want nil", got)
		}
	})

	t.Run("readings around weight anchor", func(t *testing.T) {
		anchored := append(readings,
			MeasurementReading{Kind: KindBodyTemperature, Value: 36.4, Unit: UnitCelsius, MeasuredAt: "2026-06-18T06:55:00.000Z", Source: "apple_health_export"},
			MeasurementReading{Kind: KindBodyTemperature, Value: 36.7, Unit: UnitCelsius, MeasuredAt: "2026-06-18T07:03:00.000Z", Source: "apple_health_export"},
			MeasurementReading{Kind: KindPulse, Value: 65, Unit: UnitBPM, MeasuredAt: "2026-06-18T07:11:00.000Z", Source: "apple_health_export"},
		)
		morning := SelectReadingsByWeightAnchor(anchored, PeriodMorning)
		evening := SelectReadingsByWeightAnchor(anchored, PeriodEvening)
		if got := morning[KindWeight].Value; got != 70.1 {
			t.Fatalf("morning anchor weight = %v, want 70.1", got)
		}
		if got := morning[KindBodyTemperature].Value; got != 36.7 {
			t.Fatalf("morning closest temperature = %v, want 36.7", got)
		}
		if got := evening[KindWeight].Value; got != 70.3 {
			t.Fatalf("evening anchor weight = %v, want 70.3", got)
		}
		if got := SelectReadingsByWeightAnchor(anchored[3:], PeriodMorning); len(got) != 0 {
			t.Fatalf("without weight anchor = %#v, want empty", got)
		}
	})

	t.Run("rounding resolution", func(t *testing.T) {
		cases := []struct {
			kind MeasurementKind
			in   float64
			want float64
		}{
			{KindWeight, 68.19999694824219, 68.2},
			{KindBodyTemperature, 36.66, 36.7},
			{KindBloodPressureSystolic, 120.6, 121},
			{KindPulse, 63.6, 64},
		}
		for _, tc := range cases {
			if got := RoundToMeasurementResolution(tc.in, tc.kind); got != tc.want {
				t.Errorf("round(%s, %v) = %v, want %v", tc.kind, tc.in, got, tc.want)
			}
		}
	})
}
